controllers/helpers.go

```diff
@@ -344,3 +346,16 @@ func GetMachinePoolByName(ctx context.Context, c client.Client, namespace, name
 	}
 	return m, nil
 }
+
+// ShouldDeleteIndividualResources returns false if the resource group is managed and the whole cluster is being deleted
+// meaning that we can rely on a single resource group delete operation as opposed to deleting every individual VM resource.
+func ShouldDeleteIndividualResources(ctx context.Context, clusterScope *scope.ClusterScope) bool {
+	if clusterScope.Cluster.DeletionTimestamp.IsZero() {
+		return true
+	}
+	grpSvc := groups.NewService(clusterScope)
+	managed, err := grpSvc.IsGroupManaged(ctx)
```

Comment:
I'm having a really hard time testing this... since this makes an actual API call it needs to be mocked but the mock Client we have won't work for this since it's not of type azure.Service (azure.Service doesn't have a func `IsGroupManaged`)

Reply:
Is this exposed too high up the stack? Should the services understand they should skip delete if their cluster `IsGroupManaged` rather than the controller deciding to call delete on the underlying services?

Reply:
That'd be another way of doing it... but then that would require every single service to skip the delete instead of having the logic in one central place, which might be error-prone if new services are written in the future, no? Also would mean we would call every single service no matter what instead of deleting the RG right away (not a big deal but could be a slight de-optimization).

Reply:
The other thing is that we're not currently storing the `IsGroupManaged` value anywhere in spec (AzureCluster.Spec.ResourceGroup is a `string`) unlike the vnet for which we store the Tags, so every time we check we're reaching out to Azure right now. We shouldn't GET the RG tags for each resource during delete, so doing that would require a bigger change to start storing the RG tags. I actually had started doing that initially and was looking at either adding an annotation or a Status field but then that causes the problem of refreshing the tags when a delete is triggered to make sure we are up to date. All of this should be possible, I mostly just thinking "out loud".

Reply:
Also - we don't want to skip a service deletion if the resource belongs to something that isn't an AzureCluster that is getting individually deleted. For example, a public IP should be deleted if the AzureMachine it belongs to is getting deleted in the context of a scale down, but it shouldn't be deleted if the entire cluster is getting deleted. So I actually think the controller is the right layer too put this logic, not the services.

Reply:
Let's see how this plays for a bit. I'm not convinced the controller is where this should live. However, I do agree that each service should not need to call out to Azure to find out if the resource group is managed. 

I do think there is another way around each service calling out to Azure by making once call early in the reconcile cycle and caching the result in the cluster scope struct. I would imagine this should be added to the `ClusterDescriber` interface. Perhaps, the implementation there would do the memoization.

With that said, I don't want to hold up the PR. It's definitely an improvement. Let's just think on this one some more.